hearthstone/simulator/core/player.py

```diff
@@ -282,6 +282,12 @@ def draw(self, unfreeze: Optional[bool] = True):
         number_of_cards = min(number_of_cards, self.maximum_store_size - self.store_size())
         self.extend_store(self.tavern.deck.draw(self, number_of_cards))
 
+    def draw_with_predicate(self, predicate: Callable, unfreeze: Optional[bool] = True):
```

Comment:
It looks like the draw/draw with predicate functions need to be cleaned up in general.

1. It looks like the current implementation of draw_with_predicate here does not use number_of_card.
2. It looks like there is duplication between the two functions here.  Most likely draw can be implemented in terms of draw_with_predicate with a tavern tier predicate.
3. The draw and draw_with_predicate functions on the deck seem to be redundant to me.  We could have a single function that limits by both tier and predicate as necessary (making these argument optional).
4. The draw function on the deck takes in a player, but I think it might be a better interface to directly take in the tavern tier, since this is the only piece of information used from the player.

Reply:
agreed on all points except 4 (we need the randomizer, which comes from player.tavern, and it also uses player.name)